### PR TITLE
make sysroot finding compatible with multiarch systems

### LINKED ISSUE #109994: Manually moving shared libraries to a multiarch libdir, eg, /usr/lib/x86_64-linux-gnu, breaks after 1.67
### Description

Debian deploys a multiarch schema. For Rust, `librustc_driver-*.so` and other shared libraries needed for `rustc` to run are installed to `/usr/lib/x86_64-linux-gnu` on x86_64. And `rustlib` directory is installed as `/usr/lib/rustlib`.

During Rust build, Debian just lets `x.py` install the shared libraries to `/usr/lib`, and then manually moves them to `/usr/lib/x86_64-linux-gnu`.

This works before 1.67.

But since 1.67, `rustc` will fail to find `std` crate, since `rustc` tries to find `rustlib` directory through a relative path based on `librustc_driver-*.so`. `rustc` guesses `rustlib` directory is   `/usr/lib/x86_64-linux-gnu/../lib/rustlib` aka `/usr/lib/lib/rustlib`, which does not exist.

To work around this, Ubuntu deploys a downstream [patch](https://git.launchpad.net/ubuntu/+source/rustc/tree/debian/patches/ubuntu-fix-sysroot-detection.patch?h=applied/ubuntu/lunar-devel&id=7cc6b62b69c95351eaec5c89088853aeec52d39d). But as [discussed](https://users.rust-lang.org/t/failed-to-build-rust-1-68-2-for-freedesktop-sdk-std-library-not-found/91964/8) with @bjorn3, that patch actually breaks what Rust 1.67 trying to get.

I hope Rust upstream can provide a better solution for this installation schema to work without downstream patch.

### Version it worked on

It most recently worked on: Rust 1.66.1

### Version with regression

1.67

### PR DESCRIPTION
**Pull request by onur-ozkan**
Tested on Debian 11 multiarch, worked just fine.

resolves #109994

### CODE REVIEW

File: compiler/rustc_session/src/filesearch.rs [hunk 1]
```diff
@@ -227,28 +227,30 @@ pub fn get_or_default_sysroot() -> Result<PathBuf, String> {
...
-                            Some(f) if f == "lib" => p.parent(), // first chop went for $arch, so chop again for `lib`
-                            _ => Some(p),
-                        },
-                        None => None,
-                    }
-                })
+                .and_then(|p| p.parent()) // chop off `lib`
                 .map(|s| s.to_owned())
-                .ok_or(format!(
+                .expect(&format!(
```
Comment (est31) [inline compiler/rustc_session/src/filesearch.rs:235]: IMO this should be changed to `ok_or_else` and then one can use `?` at the end.

File: compiler/rustc_session/src/filesearch.rs [hunk 2]
```diff
@@ -227,28 +227,30 @@ pub fn get_or_default_sysroot() -> Result<PathBuf, String> {
...
+        };
+
+        // On multiarch linux systems, there will be multiarch directory named
+        // with the architecture(e.g `x86_64-linux-gnu`) under the `lib` directory.
+        // Which cause us to mistakenly end up in the lib directory instead of the sysroot directory.
+        if sysroot_dir.ends_with("lib") {
+            sysroot_dir = sysroot_dir
+                .parent()
+                .map(|s| s.to_owned())
+                .expect(&format!("Could not move to parent path of {}", sysroot_dir.display()));
```
Comment (est31) [inline compiler/rustc_session/src/filesearch.rs:250]: Same here

**Review by est31 — CHANGES_REQUESTED** [review]

**Comment by jackh726** [comment]
Any way we can write a test?

**Comment by ozkanonur** [comment]
> Any way we can write a test?

The best way of testing this is having multiarch linux as CI runner(which I think we should have, just like having windows, mac and bsd runners cc @jyn514) which builds the compiler and runs the tests. Anything other than would be supper annoying hacks.

**Comment by jyn514** [comment]
cc https://github.com/rust-lang/rust/issues/102563

**Comment by jackh726** [comment]
@bors r+ rollup=never

(Given this doesn't have a test, I'm going to make sure we can bisect easily)

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-04-27/near/353314738)

@rustbot label +beta-accepted

**Pull request merged**